Theme: Add density support for surface padding

aduth opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Closes #73360

Updates theme package to implement basic support for density.

## Why?

As described in #71196, density is intended to be one of the key configurable aspects of theming. Several of the use-cases described for Boxes associated with #72784 are related to context-specific spacing adjustments (see https://github.com/WordPress/gutenberg/issues/72336#issuecomment-3499182839).

## How?

- Uses [DTCG-standard Extensions syntax](https://www.designtokens.org/tr/2025.10/format/#extensions) to provide "overrides" for semantic spacing tokens, defined using "modes" that [Terrazzo](https://terrazzo.app/) tool can understand as part of compiled CSS
- The changes here include a fix to the behavior of the plugin introduced in #72890 which handles value inlining, since it previously did not handle this type of mode overrides behavior (where mode values may also reference a primitive value in a way that needs to be appropriately inlined)

## Testing Instructions

Verify in Storybook:

1. Run `npm run storybook:dev`
2. Go to http://localhost:50240/
3. In sidebar navigation, navigate to "Design System" > "Components" > "Box" > "Default"
4. Use "Theme" button in Storybook toolbar to toggle to different densities like "Compact" and "Comfortable"
8. Observe that the padding varies based on density selection

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/15ed31a7-d0b3-4e20-8b83-bbb342cc5dc6

aduth committed:
Theme: Fix inline plugin handling of modes

aduth committed:
Theme: Add density support for surface padding

aduth committed:
UI: Update Box story demonstrating density spacing

jameskoster commented on packages/theme/src/prebuilt/css/design-tokens.css:
```diff
@@ -127,6 +127,20 @@
 	--wpds-font-size-x-small: 11px; /* Extra small font size */
 }
 
+[data-wpds-theme-provider-id][data-wpds-density='compact'] {
+	--wpds-dimension-padding-surface-large: 16px; /* Large spacing for surfaces */
+	--wpds-dimension-padding-surface-medium: 12px; /* Medium spacing for surfaces */
+	--wpds-dimension-padding-surface-small: 8px; /* Small spacing for surfaces */
+	--wpds-dimension-padding-surface-x-small: 4px; /* Extra small spacing for surfaces */
+}
+
+[data-wpds-theme-provider-id][data-wpds-density='comfortable'] {
+	--wpds-dimension-padding-surface-large: 24px; /* Large spacing for surfaces */
+	--wpds-dimension-padding-surface-medium: 20px; /* Medium spacing for surfaces */
+	--wpds-dimension-padding-surface-small: 16px; /* Small spacing for surfaces */
+	--wpds-dimension-padding-surface-x-small: 12px; /* Extra small spacing for surfaces */
+}
+
```
I don't have a strong opinion but I'm wondering if a naming convention like `xs, sm, md, lg, xl` would be better.

aduth replied:
I think we could still change it at this point if we think it'd be an improvement. For context, these are largely carried over from existing conventions like what exist in base-styles ([example](https://github.com/WordPress/gutenberg/blob/1b5de6ab1becb093d410c2cdb58314ecdb6f0582/packages/base-styles/_variables.scss#L22-L28))

jameskoster replied:
I think the shorter names are a bit easier to interpret when the scale grows. IE `4xs` is slightly more intuitive than `4-x-small`. It also makes the tokens names shorter, and a bit easier to scan in a list. So I think I lean in that direction, but only slightly. Keen to hear other thoughts cc @mattmiklic.

jameskoster commented on packages/theme/src/prebuilt/json/figma.json:
```diff
@@ -560,25 +560,33 @@
-			".": "8px"
+			".": "8px",
+			"compact": "4px",
+			"comfortable": "12px"
```
I know this is file is compiled but leaving some comments about the values here...

I think we'll want xs padding to be 4px as a default—that matches existing UIs like menus. Obviously this raises a question about what the compact value should be, but maybe it's okay for that to be the same?

From that standpoint, 4px → 12px would be quite a big jump, so maybe 8px would be good for the comfortable density.

jameskoster replied:
I suppose an alternative angle would be to add an xxs token :)

aduth replied:
I'm definitely open to any and all suggestions around specific values. My proposals should be considered largely educated guesses 🙂

jameskoster replied:
Maybe we add a smaller token then and leave this one as-is :)

aduth replied:
Just to close the loop on this, we ended up adding a `2xs` token value that's `4px` by default.

jameskoster commented on packages/theme/src/types.ts:
```diff
@@ -23,6 +23,11 @@ export interface ThemeProviderSettings {
+	/**
+	 * The density of the theme.
+	 */
+	density?: 'compact' | 'comfortable';
```
Should there be a middle-ground default, 'balanced' or similar?

jameskoster replied:
Ah, I understand there is a default, but it's not named. Feel free to resolve :D

aduth replied:
Yeah, the absence of an override is considered the default. We could have something more explicit, though I think this mirrors nicely to the way it works in the tokens, which is that the densities are "extensions" (i.e. overrides) to some base value.

mirka replied:
I think we should at least mention in the prop description that it's going to use some other default density as the default. As it looks to a consumer now, we might as well just forgotten to document whether the default is `compact` or `comfortable`.

jameskoster commented on packages/ui/src/box/stories/index.story.tsx:
```diff
@@ -53,14 +53,20 @@ export const Default: Story = {
+export const Compact: Story = {
+	...Default,
+	globals: {
+		density: 'compact',
```
I guess it could be useful to have a comfortable story too. Maybe also a story that includes all densities so you can try out the full matrix of padding/density combos.

aduth replied:
Yeah, and to your other point about Storybook configurability, this is something I had been intentionally punting, though is starting to make it difficult to demonstrate or test the differences, so might make sense to do sooner than later (i.e. maybe implement as part of this work).

jameskoster commented on packages/theme/src/prebuilt/json/figma.json:
```diff
@@ -560,25 +560,33 @@
+			"comfortable": "24px"
 		},
 		"description": "Large spacing for surfaces"
 	},
```
I think we'll probably need tokens for 32px and 48px padding. 32px is currently used in Dialogs and 48px in DataViews. 

I guess that also means we'll need a new primitive so that 48px can jump to 56px in comfortable density.

aduth committed:
Storybook: Add Theme control addon for design system stories

aduth committed:
Update theme story to demonstrate density nesting

aduth committed:
Remove Compact Box story

Now configurable through global theme configuration

jameskoster commented:
> add some global theme configuration to the Storybook toolbar.

Awesome!

> Could we maybe collaborate on a set of mappings for the initial slate of semantic surface tokens and their compact, default, and comfortable variations?

Do you mean map the tokens to real UI for a visual demo? If so, do you think it would be possible to update the Figma plugin that Marco created to import the spacing tokens? I see they're included in `figma.json` but running the importer only seems to creates color variables.

Somewhat related, I do think we'll want to support `inline` and `block` padding sooner rather than later. I was thinking about Badges as an example; they'd likely use inline padding but not block padding.

aduth commented:
> > Could we maybe collaborate on a set of mappings for the initial slate of semantic surface tokens and their compact, default, and comfortable variations?
> 
> Do you mean map the tokens to real UI for a visual demo? If so, do you think it would be possible to update the Figma plugin that Marco created to import the spacing tokens? I see they're included in `figma.json` but running the importer only seems to creates color variables.

What I had in mind was similar to https://github.com/WordPress/gutenberg/pull/72984#issuecomment-3496969844, if you had ideas for which primitives values should be referenced by each specific density + spacing token combination.

Regarding Figma, I think it's worth revisiting though I'm also holding my breath that they ship [their native support](https://www.figma.com/blog/schema-2025-design-systems-recap/#importing-and-exporting-variables) soon and it can handle this all better.

> Somewhat related, I do think we'll want to support `inline` and `block` padding sooner rather than later. I was thinking about Badges as an example; they'd likely use inline padding but not block padding.

This is supported in the component implementation itself and not in the design tokens. Do you think it needs to be in the tokens?

i.e.

```tsx
<Box padding={ { block: 'small', inline: 'medium' } } />
```

mirka commented on package.json:
```diff
@@ -235,8 +235,8 @@
 		"storybook:build": "storybook build -c ./storybook -o ./storybook/build",
-		"storybook:dev": "concurrently \"npm run dev\" \"wait-on .dev-ready && storybook dev -c ./storybook -p 50240\"",
-		"storybook:e2e:dev": "concurrently \"npm run dev\" \"wait-on .dev-ready && storybook dev -c test/storybook-playwright/storybook -p 50241\"",
+		"storybook:dev": "npm run build && storybook dev -c ./storybook -p 50240",
```
> * As far as I can tell, we don't need to be watching files manually, and can entrust that to Storybook? Maybe I'm missing something, but I tested live updates to components and it seemed to work fine.

If the basic setup hasn't changed, normal watchers like in the Storybook dev script won't be able to auto-rebuild changes in the Sass stylesheets in most wp packages. All the stylesheets are just imported in the index stylesheet and not imported by the component files that use them, so there isn't a proper dependency graph for a standard dev script to follow.

aduth replied:
Okay 👍 The current script here is busted though, because it writes `.dev-ready` before the build even finishes.

Here's some ideas:

- We could somehow shift the `.dev-ready` into the package, or expect it to provide some signal to the script when it's appropriate to write the file
- Similar to something I suggested earlier, we could do an initial build and then switch to "watch" for the Storybook. We'd probably want to add some support to `dev` to "skip initial build", also because we don't want it to clean (read: destroy) the built files

aduth replied:
Some of this could happen in a follow-up. I think the current state is much more broken than not having stylesheets recompiled on changes (a "nice to have").

mirka replied:
Weird, I just tried `npm run storybook: dev` on trunk and it's largely working for me (Storybook boots without error), aside from these warnings:

```
[1] WARN export 'default' (imported as 'FontSize') was not found in './font-sizes/font-size' (module has no exports)
[1] WARN Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
[1] WARN Deprecation Warning on line 0, column 8 of file:///Users/lena/Documents/Work/Automattic/gutenberg/storybook/global-basic.lazy.scss:0:8:
[1] WARN Sass @import rules are deprecated and will be removed in Dart Sass 3.0.0.
[1] WARN 
[1] WARN More info and automated migrator: https://sass-lang.com/d/import
[1] WARN 
[1] WARN 0 | @import "~@wordpress/base-styles/variables";
[1] WARN 
[1] WARN 
[1] WARN ../../../../storybook/global-basic.lazy.scss 1:9  root stylesheet
[1] WARN 
[1] WARN Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
[1] WARN Deprecation Warning on line 0, column 8 of file:///Users/lena/Documents/Work/Automattic/gutenberg/storybook/global-wordpress.lazy.scss:0:8:
[1] WARN Sass @import rules are deprecated and will be removed in Dart Sass 3.0.0.
[1] WARN 
[1] WARN More info and automated migrator: https://sass-lang.com/d/import
[1] WARN 
[1] WARN 0 | @import "~@wordpress/base-styles/colors";
[1] WARN 
[1] WARN 
[1] WARN ../../../../storybook/global-wordpress.lazy.scss 1:9  root stylesheet
[1] WARN 
[1] WARN
[1]  Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
[1] WARN Deprecation Warning on line 0, column 8 of file:///Users/lena/Documents/Work/Automattic/gutenberg/storybook/stories/playground/fullpage/style.lazy.scss:0:8:
[1] WARN Sass @import rules are deprecated and will be removed in Dart Sass 3.0.0.
[1] WARN 
[1] WARN More info and automated migrator: https://sass-lang.com/d/import
[1] WARN 
[1] WARN 0 | @import "~@wordpress/base-styles/colors";
[1] WARN 
[1] WARN 
[1] WARN ../../../../storybook/stories/playground/fullpage/style.lazy.scss 1:9  root stylesheet
[1] WARN 
[1] WARN Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
[1] WARN Deprecation Warning on line 1, column 8 of file:///Users/lena/Documents/Work/Automattic/gutenberg/storybook/stories/playground/fullpage/style.lazy.scss:1:8:
[1] WARN Sass @import rules are deprecated and will be removed in Dart Sass 3.0.0.
[1] WARN 
[1] WARN More info and automated migrator: https://sass-lang.com/d/import
[1] WARN 
[1] WARN 1 | @import "~@wordpress/base-styles/variables";
[1] WARN 
[1] WARN 
[1] WARN ../../../../storybook/stories/playground/fullpage/style.lazy.scss 2:9  root stylesheet
[1] WARN 
[1] WARN Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
[1] WARN Deprecation Warning on line 2, column 8 of file:///Users/lena/Documents/Work/Automattic/gutenberg/storybook/stories/playground/fullpage/style.lazy.scss:2:8:
[1] WARN Sass @import rules are deprecated and will be removed in Dart Sass 3.0.0.
[1] WARN 
[1] WARN More info and automated migrator: https://sass-lang.com/d/import
[1] WARN 
[1] WARN 2 | @import "~@wordpress/base-styles/mixins";
[1] WARN 
[1] WARN 
[1] WARN ../../../../storybook/stories/playground/fullpage/style.lazy.scss 3:9  root stylesheet
[1] WARN 
[1] WARN Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
[1] WARN Deprecation Warning on line 3, column 8 of file:///Users/lena/Documents/Work/Automattic/gutenberg/storybook/stories/playground/fullpage/style.lazy.scss:3:8:
[1] WARN Sass @import rules are deprecated and will be removed in Dart Sass 3.0.0.
[1] WARN 
[1] WARN More info and automated migrator: https://sass-lang.com/d/import
[1] WARN 
[1] WARN 3 | @import "~@wordpress/base-styles/breakpoints";
[1] WARN 
[1] WARN 
[1] WARN ../../../../storybook/stories/playground/fullpage/style.lazy.scss 4:9  root stylesheet
[1] WARN 
[1] WARN Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
[1] WARN Deprecation Warning on line 4, column 8 of file:///Users/lena/Documents/Work/Automattic/gutenberg/storybook/stories/playground/fullpage/style.lazy.scss:4:8:
[1] WARN Sass @import rules are deprecated and will be removed in Dart Sass 3.0.0.
[1] WARN 
[1] WARN More info and automated migrator: https://sass-lang.com/d/import
[1] WARN 
[1] WARN 4 | @import "~@wordpress/base-styles/animations";
[1] WARN 
[1] WARN 
[1] WARN ../../../../storybook/stories/playground/fullpage/style.lazy.scss 5:9  root stylesheet
[1] WARN 
[1] WARN Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
[1] WARN 2 repetitive deprecation warnings omitted.
[1] WARN Run in verbose mode to see all warnings.
```

(Not sure what the `FontSize` problem is, that's new to me.)

Are you seeing any errors that prevent Storybook from booting? I also checked the file history, and I couldn't really identify any changes that would fundamentally break the original wait mechanism. IIRC, we basically only need to wait for the initial `tsc --build` to complete, and that's still intact.

aduth replied:
My impression is that it's a race condition which may not always be manifesting consistently as an error.

What I can see in `trunk` which validates the potential for race conditions is that in the output of `npm run storybook:dev` , you can see transpilation messages intermixed with Storybook output, which is a strong signal that Storybook is _not_ waiting for the build to finish before starting. I believed this was the same sort of issue that prompted your changes in #72487

Example snippet showing transpilation and Storybook book happening concurrently for me on `trunk`:

```
[0]    ✔ Transpiled eslint-plugin (29ms)
[0]    ✔ Transpiled keycodes (136ms)
[0]    ✔ Transpiled react-i18n (135ms)
[0]    ✔ Transpiled date (139ms)
[0]    ✔ Transpiled a11y (139ms)
[1] @storybook/core v8.4.7
[1] 
[0]    ✔ Transpiled primitives (148ms)
[0]    ✔ Transpiled theme (158ms)
[0]    ✔ Transpiled api-fetch (159ms)
[0]    ✔ Transpiled dom (176ms)
[0]    ✔ Transpiled scripts (3ms)
```

mirka replied:
I see the same, but FWIW that's same behavior as when I first added the wait mechanism. It never waited for transpilation, or anything else in the main build script. If I understand correctly, we don't have to wait for the _entire_ build to complete, just certain "prebuild" that must be ready when the Storybook's builder (webpack) starts.

Yes it's messy, but if the Storybook is working then it's probably fine? 🙈 At least for the scope of this PR.

aduth replied:
> Yes it's messy, but if the Storybook is working then it's probably fine? 🙈 At least for the scope of this PR.

It wasn't working originally, which is what prompted the changes. 

Although testing again (via `git checkout trunk -- bin/dev.mjs .gitignore package.json` on this branch), I'm not able to reproduce the errors _as the branch exists now_.

What I think happened is that in earlier iterations of the pull request, I considered having "fancier" options in the Theme dropdown to show some previews for the color theme using [`DuotoneSwatch`](https://wordpress.github.io/gutenberg/?path=/docs/components-duotoneswatch--docs), and it was the import of that component that triggered the error.

I was able to reproduce it by adding back the swatch as a test:

```
diff --git a/storybook/addons/design-system-theme/register.tsx b/storybook/addons/design-system-theme/register.tsx
index badc085441..9f3396376e 100644
--- a/storybook/addons/design-system-theme/register.tsx
+++ b/storybook/addons/design-system-theme/register.tsx
@@ -11,2 +11,3 @@ import {
 } from '@storybook/components';
+import { DuotoneSwatch } from '@wordpress/components';
 
@@ -83,2 +84,3 @@ const ThemeTool = () => {
 				<MirrorIcon />
+				<DuotoneSwatch />
 				Theme
``` 

At that point running `npm run storybook:dev` I see:

```
[1] ✘ [ERROR] Could not resolve "@wordpress/components"
[1] 
[1]     storybook/addons/design-system-theme/register.tsx:12:30:
[1]       12 │ import { DuotoneSwatch } from '@wordpress/components';
[1]          ╵                               ~~~~~~~~~~~~~~~~~~~~~~~
[1] 
[1]   The module "./build-module/index.js" was not found on the file system:
[1] 
[1]     node_modules/@wordpress/components/package.json:30:13:
[1]       30 │       "import": "./build-module/index.js",
[1]          ╵                 ~~~~~~~~~~~~~~~~~~~~~~~~~
[1] 
[1]   You can mark the path "@wordpress/components" as external to exclude it from the bundle, which will remove this error and leave the unresolved path in the bundle.
[1] 
[1] Error: Build failed with 1 error:
[1] storybook/addons/design-system-theme/register.tsx:12:30: ERROR: Could not resolve "@wordpress/components"
```

For the purpose of this pull request as is and to avoid holding this up, I can revert these changes for now, but I expect it's likely the issue will manifest again in the future.

mirka replied:
I see, yeah that's bad, thanks for posting that repro. I agree we'll need some improvements like you suggested in https://github.com/WordPress/gutenberg/pull/73215#discussion_r2527916758, sooner or later.

aduth committed:
Revert changes to Storybook build with dev-ready indicator

See: https://github.com/WordPress/gutenberg/pull/73215/files#r2527844132

aduth committed:
Update size options to extend to baseline 4px spacing

mirka commented on packages/theme/src/stories/index.story.tsx:
```diff
@@ -198,10 +198,10 @@ export const WithPicker: StoryObj< typeof ThemeProvider > = {
 };
 
-const NestingDebug = ( { bg = '', primary = '' } ) => (
+const NestingDebug = ( { bg = '', primary = '', density = '' } ) => (
```
The density inheritance isn't working yet, right? We need to add handling in `useThemeProviderStyles()`.

aduth replied:
I believe the inheritance is working correctly, since the `[data-wpds-density]` from an ancestor will take effect automatically unless another non-`undefined` `density=` is provided to a descendent `ThemeProvider`. 

You can check in the updated Storybook story (after fixes in recent commits): http://localhost:50240/?path=/story/design-system-theme-provider--nesting-and-inheriting

The first item has a padding of `16px` (default), but all descendents have compact padding (`12px`).

However, related to [your other comment about defaults](https://github.com/WordPress/gutenberg/pull/73215#discussion_r2543117563), one thing which _isn't_ supported here is explicitly resetting density back to default. We could add an explicit `"default"` option, and then maybe handle this either through `useThemeProviderStyles` (copy the root styles) or somehow in the CSS output.

aduth replied:
The changes in 94848d3d18 should address the need to be able to reset back to an explicit default density (if inheriting would otherwise inherit a non-default value). It also expands the prop description to describe the behavior better.

mirka commented on packages/theme/src/stories/index.story.tsx:
```diff
@@ -198,10 +198,10 @@ export const WithPicker: StoryObj< typeof ThemeProvider > = {
 	<div
 		style={ {
-			padding: '0.25rem',
+			padding: 'var(--wpds-dimension-padding-surface-small)',
```
Probably want to update the faux button ("primary"/"Neutral") paddings in this component to use the tokens too.

aduth replied:
I'd agree but only if we don't think of them as buttons 😅 We might end up having semantic tokens for buttons, but I don't think they'd be "surface" tokens (similar to how we differentiate `bg-surface` and `bg-interactive`). But if we think of them as fake badges, surface could be appropriate.

aduth replied:
Updated to use padding tokens in 56672d58da. I think this also does a better job of demonstrating the effect of density in the stories:

<img width="963" height="321" alt="image" src="https://github.com/user-attachments/assets/dd3d88c2-aa42-4bec-978b-4f42043cf478" />

mirka commented on storybook/addons/design-system-theme/register.tsx:
```diff
@@ -0,0 +1,99 @@
+	addons.add( `${ ADDON_ID }/tool`, {
+		type: types.TOOL,
+		title: 'Design System Theme',
+		match: ( { storyId, viewMode } ) =>
+			!! storyId?.startsWith( 'design-system-' ) &&
+			( [ 'story', 'docs' ] as any[] ).includes( viewMode ),
```
If it's easy, we may want to hide the addon from the `ThemeProvider` stories since the theme values are hardcoded there.

aduth replied:
I'll test, but can probably...

```suggestion
		match: ( { storyId, viewMode } ) =>
			!! storyId?.startsWith( 'design-system-components-' ) &&
			( [ 'story', 'docs' ] as any[] ).includes( viewMode ),
```

aduth replied:
Updated in c9c8985c57

mirka commented on storybook/addons/design-system-theme/register.tsx:
```diff
@@ -0,0 +1,99 @@
+
+const ThemeTool = () => {
+	return (
+		<WithTooltip
+			placement="top"
+			trigger="click"
+			closeOnOutsideClick
+			tooltip={
```
I just noticed how inaccessible these toolbar controls are, including the standard ones we already have 😂 So weird… oh well.

aduth replied:
> I just noticed how inaccessible these toolbar controls are, including the standard ones we already have 😂 So weird… oh well.

Yeah, this was something I noticed as well 😭 Not sure if it's any better in more recent versions of Storybook.

mirka commented on storybook/addons/design-system-theme/register.tsx:
```diff
@@ -0,0 +1,99 @@
+				</>
+			}
+		>
+			<IconButton key="density" title="Design System Density" active>
```
```suggestion
			<IconButton title="Design System" active>
```

Also not sure what the `active` prop is here 🤔 And I could not find the documentation for the `@storybook/components` package within a few minutes 🙃

aduth replied:
> Also not sure what the `active` prop is here 🤔 And I could not find the documentation for the `@storybook/components` package within a few minutes 🙃

Short answer is: `active` makes the toolbar item appear blue instead of a more disabled-looking default gray.

<img width="57" height="34" alt="image" src="https://github.com/user-attachments/assets/7f137216-4099-4e12-89bb-8afa7f365be0" />

As far as documentation... I'm struggling to find it as well 😅 Looking back through my browser history, I found [this Chromatic component library](https://www.chromatic.com/library?appId=5ccbc373887ca40020446347). Today I was also able to track down [the "Storybook Design System" Storybook](https://storybook-design-system.netlify.app/?path=/docs/intro--page) (unclear if official). These include some of the components being used here (e.g. `TooltipMessage`), but not `IconButton` for some reason.

aduth replied:
Regarding the suggestions, I'll take another look at this. I think some of these may have been sourced from reference documentation ([example](https://storybook.js.org/docs/essentials/toolbars-and-globals#updating-globals-from-within-an-addon)), but it's not entirely clear if they're needed (particularly `key`, not as confident yet on `title`).

aduth replied:
Went with your suggestion in 80cbbeabf2 , adapted slightly to include "Theme" (as visibly shown).

aduth committed:
Use updated name for small padding token

aduth committed:
Ensure default CSS included in Storybook story

aduth committed:
Add explicit default density option

Allow resetting if unset density would otherwise inherit from a non-default density ancestor

aduth committed:
Limit Storybook theming options to design system components

Avoid interfering with theming shown in ThemeProvider stories

aduth committed:
Improve title of IconButton for Theme toolbar item

See: https://github.com/WordPress/gutenberg/pull/73215#discussion_r2543188865
Co-Authored-By: Lena Morita <555336+mirka@users.noreply.github.com>

aduth committed:
Use padding tokens for primary, neutral color demonstration

mirka commented on packages/theme/src/types.ts:
```diff
@@ -23,6 +23,15 @@ export interface ThemeProviderSettings {
 		bg?: string;
 	};
+
+	/**
+	 * The density of the theme. If left unspecified, the theme inherits from
+	 * the density of the closest `ThemeProvider`, or uses the default density
+	 * if there is no inherited density.
+	 *
+	 * @default undefined
+	 */
+	density?: undefined | 'default' | 'compact' | 'comfortable';
```
Just a thought, no strong opinion, but would something like this be more intuitive? Personally I think it would be easier to understand when skimming the docs, since it's closer to the CSS mental model.

```suggestion
	/**
	 * The density of the theme. If left unspecified, the theme inherits from
	 * the density of the closest `ThemeProvider`, or uses the default density
	 * if there is no inherited density.
	 *
	 * @default 'inherit'
	 */
	density?: 'inherit' | 'default' | 'compact' | 'comfortable';
```

aduth replied:
Hm, I think there's some advantages to that, particularly around explicitness. However, on the whole, I don't know that it's a change I want to make now:

- The idea of using `undefined` (absence of value) as interpreted to inherit from the parent is consistent with how we're treating color customization above ([source](https://github.com/WordPress/gutenberg/blob/01136d19b0985d12ff42c3cb7329ea421a583d28/packages/theme/src/types.ts#L11-L24))
- While not really a blocker in itself, doing this incurs some internal technical complexity because we need to normalize the `'inherit'` value _back to_ `undefined` in order to effect the inherit behavior by ensuring that `data-wpds-density` isn't assigned ([source](https://github.com/WordPress/gutenberg/blob/56672d58dab126c9e9a1c3cd5a6d1cc48c5f9df6/packages/theme/src/theme-provider.tsx#L80)) because, unlike `[data-wpds-density="compact"]`, etc., we don't have an explicit selector for the inherited behavior (there is no `[data-wpds-density="inherit"]`).
- The suggestion feels like a logical step in part because of how I've gone out of my way to be explicit here, though in retrospect I'm not fully convinced it was necessary to be so explicit (i.e. including `@default undefined` and including `undefined` in the type union for `density`).

mirka replied:
That's fair. Maybe it looks weird to me because all our "physical" components do `@default 'default'`, and this is the first time we're doing props with inheritance.

mirka reviewed (approved):
Works well! 🚀
